### ISSUE TITLE
[Clang] [KCFI] why does KCFI make no sense for hook function? 

### ISSUE THREAD

**Issue by Zhenhang1213** [llvm/llvm-project#106344]
why does KCFI make no sense for hook function?  I found that the generated assembly instructions are the same in 
[godbolt ](https://godbolt.org/z/bhcoEEGPb)

**Comment by shafik** [comment]
CC @samitolvanen

**Comment by samitolvanen** [comment]
This seems to be because the function declarations don't have prototypes. Adding `-Wstrict-prototypes` to the command line results in:
```
<source>:5:49: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
    5 | typedef VOS_UINT32 (*FEIBRK_XFLOWBASE_INIT_FUNC)();
      |                                                 ^
      |                                                  void
<source>:6:50: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
    6 | typedef VOS_VOID (*FEIBRK_XFLOWBASE_DESTORY_FUNC)();
      |                                                  ^
      |                                                   void
...
```
Adding missing `void`s to the function declarations results in KCFI checks being emitted: https://godbolt.org/z/oT6v6cdWW

**Comment by Zhenhang1213** [comment]
> This seems to be because the function declarations don't have prototypes. Adding `-Wstrict-prototypes` to the command line results in:
> 
> ```
> <source>:5:49: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
>     5 | typedef VOS_UINT32 (*FEIBRK_XFLOWBASE_INIT_FUNC)();
>       |                                                 ^
>       |                                                  void
> <source>:6:50: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
>     6 | typedef VOS_VOID (*FEIBRK_XFLOWBASE_DESTORY_FUNC)();
>       |                                                  ^
>       |                                                   void
> ...
> ```
> 
> Adding missing `void`s to the function declarations results in KCFI checks being emitted: https://godbolt.org/z/oT6v6cdWW

Thanks，but this should be a bug.  Is it possible to make a compatibility improvement?

**Comment by samitolvanen** [comment]
Agreed, this would be useful case to support. Basically, we'll need to get a `QualType` for pointers without prototypes in `CodeGenFunction::EmitKCFIOperandBundle` so we can pass the type to `CodeGenModule::CreateKCFITypeId` and add the operand bundle to these calls. I'll try to find some time to look into this in the next couple of weeks, but I'm also happy to review patches!
